geomstats/solvers.py

```diff
@@ -0,0 +1,723 @@
+import logging
+from abc import ABCMeta, abstractmethod
+
+# TODO: numpy needs to be removed
+import numpy as np
+from scipy.integrate import odeint, solve_bvp
+from scipy.optimize import minimize
+
+import geomstats.backend as gs
+from geomstats.integrator import integrate
+
+
+class LogSolver(metaclass=ABCMeta):
+    @abstractmethod
+    def solve(self, metric, point, base_point):
+        pass
+
+
+class GeomstatsLogSolver(LogSolver):
+    def __init__(self, method="L-BFGS-B", max_iter=25, verbose=False, atol=gs.atol):
+        super().__init__()
+
+        self.method = method
+        self.max_iter = max_iter
+        self.verbose = verbose
+        self.atol = atol
+
+        self.res_ = None
+
+    def solve(self, metric, point, base_point):
+        """Compute logarithm map associated to the affine connection.
+
+        Solve the boundary value problem associated to the geodesic equation
+        using the Christoffel symbols and conjugate gradient descent.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., dim]
+            Point on the manifold.
+        base_point : array-like, shape=[..., dim]
+            Point on the manifold.
+        n_steps : int
+            Number of discrete time steps to take in the integration.
+            Optional, default: N_STEPS.
+        step : str, {'euler', 'rk4'}
+            Numerical scheme to use for integration.
+            Optional, default: 'euler'.
+        max_iter
+        verbose
+        tol
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., dim]
+            Tangent vector at the base point.
+        """
+        max_shape = point.shape
+        if len(point.shape) <= len(base_point.shape):
+            max_shape = base_point.shape
+
+        def objective(velocity):
```

Comment:
This could be a method of the solver class. This would answer @Jules-Deschamps's which to enable changing the objective.